src/poll/responser.rs

```diff
@@ -0,0 +1,666 @@
+                        .with_common(common)
+                        .with_sge(sge);
+                    if let Err(e) = device.push(desc_builder) {
+                        eprintln!("Failed to push ack/nack packet: {:?}", e);
+                    }
+                }
+                Ok(RespCommand::ReadResponse(resp)) => {
+                    // send read response to device
+                    let dpqn = resp.desc.common.dqpn;
+                    let (dst_ip, common) = match qp_table.read().unwrap().get(&dpqn) {
+                        Some(qp) => {
+                            let dst_ip = qp.dqp_ip;
+                            let mut common = ToCardWorkRbDescCommon {
+                                total_len: resp.desc.len,
+                                rkey: resp.desc.rkey,
+                                raddr: resp.desc.raddr,
+                                dqp_ip: qp.dqp_ip,
+                                dqpn: dpqn,
+                                mac_addr: qp.mac_addr,
+                                pmtu: qp.pmtu.clone(),
+                                flags: 0,
+                                qp_type: qp.qp_type.clone(),
+                                psn: 0,
+                            };
+                            let mut inner = qp.inner.lock().unwrap();
+                            let psn = inner.send_psn;
+                            let first_pkt_max_len =
+                                u64::from(&qp.pmtu) - (resp.desc.raddr % (u64::from(&qp.pmtu) - 1));
+                            let first_pkt_len = resp.desc.len.min(first_pkt_max_len as u32);
+                            let pkt_cnt = 1
+                                + (resp.desc.len - first_pkt_len)
+                                    .div_ceil(u64::from(&qp.pmtu) as u32);
+                            inner.send_psn += pkt_cnt;
+                            common.psn = psn;
+                            (dst_ip, common)
+                        }
+                        None => {
+                            eprintln!("Failed to get QP from QP table: {:?}", dpqn);
+                            continue;
+                        }
+                    };
+
+                    let sge = Sge {
+                        addr: resp.desc.laddr,
+                        len: resp.desc.len,
+                        key: resp.desc.lkey,
+                    };
+                    let desc_builder = ToCardWorkRbDescBuilder::new_read_resp()
+                        .with_common(common)
+                        .with_sge(sge);
+                    if let Err(e) = device.push(desc_builder) {
+                        eprintln!("Failed to push read response: {:?}", e);
+                    }
+                }
+                Err(_) => {
+                    // The only error is pipe broken, so just exit the thread
+                    return;
+                }
+            }
+        }
+    }
+}
+
+type Slot = [u8];
```

Comment:
should we make Slot an owned type instead of reference?

Reply:
All the slots are references to [u8], you can refer to `AcknowledgeBuffer::alloc`. But it seems it will be confusing to name the `slot` as `[u8]` rather than `&'static mut Slot`. I will update it soon.